File: passport-elysia/scripts/migrate.ts
#!/usr/bin/env bun
import { sql } from '../src/db';
import { readdir } from 'node:fs/promises';
import { join } from 'node:path';

async function createMigrationsTable() {
  await sql`
    CREATE TABLE IF NOT EXISTS schema_migrations (
      version TEXT PRIMARY KEY,
      executed_at TIMESTAMPTZ NOT NULL DEFAULT NOW()
    )
  `;
}

async function getExecutedMigrations(): Promise<Set<string>> {
  const rows = await sql`SELECT version FROM schema_migrations`;
  return new Set(rows.map(row => row.version));
}

async function runMigration(filePath: string, version: string) {
  const migrationSql = await Bun.file(filePath).text();
  
  // Use advisory lock to prevent concurrent migrations
  await sql`SELECT pg_advisory_lock(12345)`;
  
  try {
    // Begin transaction
    await sql.begin(async sql => {
      // Execute migration
      await sql.unsafe(migrationSql);
      
      // Record migration
      await sql`
        INSERT INTO schema_migrations (version)
        VALUES (${version})
      `;
    });
    
    console.log(`✅ Migration ${version} executed successfully`);
  } catch (error) {
    console.error(`❌ Migration ${version} failed:`, error);
    throw error;
  } finally {
    // Release advisory lock
    await sql`SELECT pg_advisory_unlock(12345)`;
  }
}

async function migrate() {
  console.log('🔄 Running database migrations...\n');
  
  try {
    // Create migrations table if not exists
    await createMigrationsTable();
    
    // Get list of executed migrations
    const executedMigrations = await getExecutedMigrations();
    
    // Read all migration files
    const migrationsDir = join(import.meta.dir, '../db/migrations');
    const files = await readdir(migrationsDir);
    const migrationFiles = files
      .filter(f => f.endsWith('.sql'))
      .sort(); // Ensure migrations run in order
    
    let migrationsRun = 0;
    
    for (const file of migrationFiles) {
      const version = file.replace('.sql', '');
      
      if (executedMigrations.has(version)) {
        console.log(`⏭️  Skipping ${version} (already executed)`);
        continue;
      }
      
      const filePath = join(migrationsDir, file);
      await runMigration(filePath, version);
      migrationsRun++;
    }
    
    if (migrationsRun === 0) {
      console.log('✨ All migrations are up to date');
    } else {
      console.log(`\n✨ Successfully ran ${migrationsRun} migration(s)`);
    }
    
    await sql.end();
    process.exit(0);
  } catch (error) {
    console.error('\n❌ Migration failed:', error);
    await sql.end();
    process.exit(1);
  }
}

// Run migrations
migrate();

File: go-passport/internal/config/migrate.go
package config

import (
	"context"
	"database/sql"
	"fmt"
	"io/ioutil"
	"path/filepath"
	"sort"
	"strings"
)

type Migrator struct {
	db *Database
}

func NewMigrator(db *Database) *Migrator {
	return &Migrator{db: db}
}

func (m *Migrator) Run(ctx context.Context, migrationsDir string) error {
	// Create migrations table if not exists
	if err := m.createMigrationsTable(ctx); err != nil {
		return fmt.Errorf("failed to create migrations table: %w", err)
	}

	// Get list of migration files
	files, err := ioutil.ReadDir(migrationsDir)
	if err != nil {
		return fmt.Errorf("failed to read migrations directory: %w", err)
	}

	// Sort migration files by name
	var migrationFiles []string
	for _, file := range files {
		if !file.IsDir() && strings.HasSuffix(file.Name(), ".sql") {
			migrationFiles = append(migrationFiles, file.Name())
		}
	}
	sort.Strings(migrationFiles)

	// Run each migration
	for _, file := range migrationFiles {
		if err := m.runMigration(ctx, filepath.Join(migrationsDir, file)); err != nil {
			return fmt.Errorf("failed to run migration %s: %w", file, err)
		}
	}

	return nil
}

func (m *Migrator) createMigrationsTable(ctx context.Context) error {
	query := `
		CREATE TABLE IF NOT EXISTS schema_migrations (
			version VARCHAR(255) PRIMARY KEY,
			applied_at TIMESTAMPTZ NOT NULL DEFAULT CURRENT_TIMESTAMP
		)`
	
	_, err := m.db.ExecContext(ctx, query)
	return err
}

func (m *Migrator) runMigration(ctx context.Context, filePath string) error {
	// Extract version from filename
	version := filepath.Base(filePath)
	
	// Check if migration has already been applied
	var count int
	err := m.db.QueryRowContext(ctx, 
		"SELECT COUNT(*) FROM schema_migrations WHERE version = $1", 
		version,
	).Scan(&count)
	
	if err != nil {
		return fmt.Errorf("failed to check migration status: %w", err)
	}
	
	if count > 0 {
		// Migration already applied
		return nil
	}
	
	// Read migration file
	content, err := ioutil.ReadFile(filePath)
	if err != nil {
		return fmt.Errorf("failed to read migration file: %w", err)
	}
	
	// Execute migration in a transaction
	return m.db.WithTransaction(ctx, func(tx *sql.Tx) error {
		// Execute migration SQL
		if _, err := tx.ExecContext(ctx, string(content)); err != nil {
			return fmt.Errorf("failed to execute migration: %w", err)
		}
		
		// Record migration as applied
		if _, err := tx.ExecContext(ctx,
			"INSERT INTO schema_migrations (version) VALUES ($1)",
			version,
		); err != nil {
			return fmt.Errorf("failed to record migration: %w", err)
		}
		
		fmt.Printf("Applied migration: %s\n", version)
		return nil
	})
}